### PR TITLE
GH-25 run the chatsample app

### PR DESCRIPTION
**Pull request by davidalpert**
This PR will capture my experiments in getting the `chatsample` up and running as a demonstration of how to initialize and use the `signalr` package.

This PR contains no changes to the `signalr` package and is entirely focused on the `chatsample` demo app.

### CODE REVIEW

**Comment by davidalpert** [comment]
currently, I can run the chatsample app with `make run-chatsample` and when I navigate to http://localhost:8086 as stdout suggests I see a web form but there are client-side console errors:
image

and the server is registering a 501:
```
~/go/src/github.com/phillippseith/signalr: make run-chatsample
Serving public content from the embedded filesystem
Listening for websocket connections on http://localhost:8086
200 GET / 3.590884ms
200 GET /js/signalr.js 461.281µs
200 GET /js/signalr.js.map 370.403µs
000 POST /chat/negotiate 310.765µs
501 GET /chat?id=x0Ak1kB7Wq5YtB9RLgCWsw== 93.572µs
000 POST /chat/negotiate 21.452µs
```

**Comment by davidalpert** [comment]
and the unit tests have failed because the github actions are running with go 1.15 (which doesn't support the `embed` package) but the `go.mod` file is pinned to go 1.16 so this will also have to be resolved.

**Comment by philippseith** [comment]
> currently, I can run the chatsample app with `make run-chatsample` and when I navigate to http://localhost:8086 as stdout suggests I see a web form but there are client-side console errors:
> image
> 
> and the server is registering a 501:
> 
> ```
> ~/go/src/github.com/phillippseith/signalr: make run-chatsample
> Serving public content from the embedded filesystem
> Listening for websocket connections on http://localhost:8086
> 200 GET / 3.590884ms
> 200 GET /js/signalr.js 461.281µs
> 200 GET /js/signalr.js.map 370.403µs
> 000 POST /chat/negotiate 310.765µs
> 501 GET /chat?id=x0Ak1kB7Wq5YtB9RLgCWsw== 93.572µs
> 000 POST /chat/negotiate 21.452µs
> ```

I will look into that. Honestly, I didn't run the chatsample for a long time and only relied on the unit tests. The tests are using the @microsoft/signalr typescript package and I am not sure how old the js code in the sample is. 
But anyway, the server should also run with old client code.

**Comment by philippseith** [comment]
The problem with run-chatsample is caused by the middleware. Without, it works. 
If you change the supported server transport to SSE (add the signalr.HTTPTransports("ServerSentEvents") option in runHTTPServer) you will get a panic because the wrapResponseWriter is no http.Flusher (which can be fixed easily). After adding the Flusher implementation, the middleware works with the SSE transport.
I don't know why nhooyr.io/websocket doesn't play well with the middleware (at least it doesn't panic :-)). 
So either you go with SSE (which won't hold you off understanding what's going on) or you use WireShark (that's what I did)

BTW: I changed to go 1.16 in the GitHub actions

**Comment by davidalpert** [comment]
thank you for your response! 

before seeing that you had updated the actions to go 1.16 locally I updated the signalr libraries and then the sample started working but ran into the same painc at the middleware not implementing `http.Flusher`

that is indeed a simple fix.

**Comment by davidalpert** [comment]
and the chatsample is alive! 👏

**Comment by davidalpert** [comment]
looks like this PR now failed only on uploading code coverage results

File: README.md [hunk 1]
```diff
@@ -16,3 +16,203 @@ This repository contains an implementation of a SignalR server and a SignalR cli
...
+    fmt.Printf("Listening for websocket connections on http://%s\n", address)
+    if err := http.ListenAndServe(address, router); err != nil {
+        log.Fatal("ListenAndServe:", err)
+    }
+}
+```
+
+### Client side
+
+#### Grab copies of the signalr scripts
```
Comment (philippseith) [inline README.md:130]: Many web devs are using package managers, so something like `npm install @microsoft/signalr` should be mentioned.

File: README.md [hunk 2]
```diff
@@ -16,3 +16,203 @@ This repository contains an implementation of a SignalR server and a SignalR cli
...
+func runHTTPServer() {
+    address := 'localhost:8080'
+    
+    // create an instance of your hub
+    hub := AppHub{}
+	
+    // build a signalr.Server using your hub
+    server, _ := signalr.NewServer(context.TODO(), signalr.SimpleHubFactory(hub))
+    
+    // configure global signalr options
```
Comment (philippseith) [inline README.md]: These options are not global, they need to be passed to NewServer(). See [Functional options for friendly APIs](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) for the pattern.

**Comment by philippseith** [comment]
> looks like this PR now failed only on uploading code coverage results

You might need a CODECOV_TOKEN secret in your repo to let the upload succeed. I additionally changed the action workflow to ignore a failing upload

File: README.md (hunk 1 above)
Comment (davidalpert) [reply README.md:130]: good idea, I'll fix that.

File: README.md (hunk 2 above)
Comment (davidalpert) [reply README.md]: thank you, I will review that.

Comment (davidalpert) [reply README.md]: Ah, I see. I misread this in the `chatsample` `runHTTPServer` function. Will fix this up.

File: README.md (hunk 1 above)
Comment (davidalpert) [reply README.md:130]: replaced that section with a table showing the commands for npm, yarn, and libman

File: README.md (hunk 2 above)
Comment (davidalpert) [reply README.md]: fixed

**Comment by davidalpert** [comment]
hmmm... the [Usage](https://github.com/codecov/codecov-action#usage) section of the `Codecov GitHub Action` readme says that
> For public repositories no token is needed

🤔

**Comment by philippseith** [comment]
> hmmm... the [Usage](https://github.com/codecov/codecov-action#usage) section of the `Codecov GitHub Action` readme says that
> 
> > For public repositories no token is needed
> 
> 🤔

I tried it without a token here 7bd3564c1d641a696489d43c4d42847e68decfc8, but it didn't work

**Pull request merged**